proof/invariant-abstract/TcbAcc_AI.thy

```diff
@@ -1824,16 +1816,367 @@ lemma get_tcb_ko_atD:
+  apply (wpsimp wp: set_object_valid_objs)
+  apply (clarsimp simp: obj_at_def dest!: get_tcb_SomeD)
+  apply (erule (1) pspace_valid_objsE)
+  apply (clarsimp simp: valid_obj_def valid_tcb_def)
+  apply (drule (1) bspec)
+  by (auto simp: tcb_cap_cases_def)
+
+lemma thread_set_domain_if_live_then_nonz_cap[wp]:
+  "thread_set_domain t d \<lbrace>if_live_then_nonz_cap\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def
+  apply wpsimp
+  apply (auto simp: if_live_then_nonz_cap_def obj_at_def tcb_cap_cases_def dest!: get_tcb_SomeD)
+  done
+
+lemma thread_set_domain_zombies_final[wp]:
+  "thread_set_domain t d \<lbrace>zombies_final\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def
+  by (wpsimp simp: tcb_cap_cases_def) auto
+
+lemma thread_set_domain_refs_of[wp]:
+  "thread_set_domain t d \<lbrace>\<lambda>s. P (state_refs_of s)\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def set_object_def get_object_def
+  apply (wpsimp simp: state_refs_of_def)
+  apply (erule rsubst[of P])
+  apply (rule ext)
+  apply (clarsimp split: option.splits dest!: get_tcb_SomeD)
+  done
+
+lemma thread_set_domain_hyp_refs_of[wp]:
+  "thread_set_domain t d \<lbrace>\<lambda>s. P (state_hyp_refs_of s)\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def set_object_def get_object_def
+  supply fun_upd_apply [simp del]
+  apply wpsimp
+  apply (clarsimp elim!: rsubst[where P=P] dest!: get_tcb_SomeD)
+  apply (subst state_hyp_refs_of_tcb_domain_update; auto simp: get_tcb_def)
+  done
+
+lemma thread_set_domain_valid_idle[wp]:
+  "thread_set_domain t d \<lbrace>valid_idle\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def
+  apply wpsimp
+  apply (auto simp: obj_at_def valid_idle_def pred_tcb_at_def dest!: get_tcb_SomeD)
+  done
+
+lemma thread_set_domain_if_unsafe_then_cap[wp]:
+  "thread_set_domain t d \<lbrace>if_unsafe_then_cap\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def
+  by (wpsimp simp: tcb_cap_cases_def) auto
+
+lemma thread_set_domain_valid_irq_node[wp]:
+  "thread_set_domain t d \<lbrace>valid_irq_node\<rbrace>"
+  apply (wpsimp simp: valid_irq_node_def wp: hoare_vcg_all_lift)
+   apply (rule hoare_lift_Pf[where f="interrupt_irq_node"]; wp cap_table_at_typ_at)
   apply simp
   done
 
+lemma thread_set_domain_valid_irq_handlers[wp]:
+  "thread_set_domain t d \<lbrace>valid_irq_handlers\<rbrace>"
+  apply (wpsimp simp: valid_irq_handlers_def irq_issued_def)
+  apply (rule hoare_lift_Pf[where f="caps_of_state"]; wp)
+  done
+
+lemma thread_set_domain_valid_arch_caps[wp]:
+  "thread_set_domain t d \<lbrace>valid_arch_caps\<rbrace>"
+  unfolding thread_set_domain_def
+  by (wpsimp wp: thread_set_arch_caps_trivial simp: tcb_cap_cases_def) auto
+
+lemma thread_set_domain_valid_ioports[wp]:
+  "thread_set_domain t d \<lbrace>valid_ioports\<rbrace>"
+  unfolding thread_set_domain_def
+  by (wpsimp wp: valid_ioports_lift thread_set_caps_of_state_trivial2 simp: tcb_cap_cases_def)
```

Comment:
This is now rebased over `valid_ioports` I think -- should we double check what is happening with this one?

Reply:
You should not be seeing that in a non-arch-specific file

Reply:
`l4v/proof/invariant-abstract]# ag valid_iop *.thy` should return nothing

Reply:
Sorry, I just discovered this independently, this a remnant of my original changes I didn't catch during the rebase. I've just pushed a new commit that fixes these proofs.

Reply:
I looked at that commit... to the untrained eye, it looks like quite a lot got pulled into `TcbAcc_AI_valid_ipc_buffer_cap_0`. Even though it's not like we're pulling things into Arch or anything, so it's fine to leave it, I'm still curious why all that got dragged in.

Reply:
The `valid_ioports` changes added `thread_set_valid_arch_state` to `TcbAcc_AI_valid_ipc_buffer_cap_0`, and this section of proofs uses that (and then needs to refer to the same `'state_ext` as in that locale). Technically it's just `thread_set_domain_valid_arch_state`, `thread_set_priority_valid_arch_state`, `thread_set_domain_invs`, and `thread_set_priority_invs` that need to be in the locale, but I thought I might as well keep things simple and put the whole section into it instead of dropping in and out of the locale multiple times.

Reply:
Yeah, that's fine then. I think I'm trying to minimise what I put in to these locales since I'm in the process of arch-split itself and don't know what will or will not end up as a circular dependency. Don't think you need to worry about that.